core/variable_operations.cpp

```diff
@@ -264,36 +264,110 @@ Variable reverse(Variable var, const Dim dim) {
 /// Return a deep copy of a Variable or of a VariableView.
 Variable copy(const VariableConstView &var) { return Variable(var); }
 
-VariableView nan_to_num(const VariableConstView &var,
-                        const VariableConstView &replacement,
-                        const VariableView &out) {
-  using std::isnan;
+namespace {
+template <typename Op>
+VariableView replace_in_place(const VariableConstView &var,
+                              const VariableConstView &replacement, Op op,
+                              const VariableView &out) {
+
   transform_in_place<std::tuple<double, float>>(
       out, var, replacement,
       scipp::overloaded{
           transform_flags::expect_all_or_none_have_variance,
-          [](auto &a, const auto &b, const auto &repl) {
-            a = isnan(b) ? repl : b;
-          },
+          [&op](auto &a, const auto &b, const auto &repl) { op(a, b, repl); },
           [](units::Unit &a, const units::Unit &b, const units::Unit &repl) {
             expect::equals(b, repl);
             a = b;
           }});
   return out;
 }
 
-Variable nan_to_num(const VariableConstView &var,
-                    const VariableConstView &replacement) {
-  using std::isnan;
+template <typename Op>
+Variable replace(const VariableConstView &var,
+                 const VariableConstView &replacement, Op op) {
   return transform<std::tuple<double, float>>(
       var, replacement,
-      overloaded{
-          transform_flags::expect_all_or_none_have_variance,
-          [](const auto &x, const auto &repl) { return isnan(x) ? repl : x; },
-          [](const units::Unit &x, const units::Unit &repl) {
-            expect::equals(x, repl);
-            return x;
-          }});
+      overloaded{transform_flags::expect_all_or_none_have_variance,
+                 [&op](const auto &x, const auto &repl) { return op(x, repl); },
+                 [](const units::Unit &x, const units::Unit &repl) {
+                   expect::equals(x, repl);
+                   return x;
+                 }});
+}
+} // namespace
+
+VariableView nan_to_num(const VariableConstView &var,
+                        const VariableConstView &replacement,
+                        const VariableView &out) {
+
+  auto op = [](auto &a, const auto &b, const auto &repl) {
+    if constexpr (is_ValueAndVariance_v<std::decay_t<decltype(b)>>)
+      a = isnan(b) ? repl : b;
+    else
+      a = std::isnan(b) ? repl : b;
```

Comment:
```suggestion
    using std::isnan;
    a = isnan(b) ? repl : b;
```
?

Reply:
Yes, I deliberately split so that we did not have ADL at a point when I intended to solve this a slightly different way. Given the single lambda passed to `replace` yes this can now be collapsed.